src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,406 @@
+                }
+
+                if ((tupleConversion.Conversion.Kind == ConversionKind.ImplicitTupleLiteral || tupleConversion.Conversion.Kind == ConversionKind.Identity)
+                    && IsTupleExpression(tupleConversion.Operand.Kind))
+                {
+                    var coreArgument = WithoutImplicitNullableConversions(((BoundTupleExpression)tupleConversion.Operand).Arguments[i]);
+                    return MakeTemp(coreArgument, operators[i], initialEffectsAndTemps, isRight);
+                }
+            }
+
+            Debug.Assert(tuple.Type.IsTupleType);
+
+            // Example:
+            // t == GetTuple();
+            // t == ((byte, byte)) (1, 2);
+            // t == ((short, short))((int, int))(1L, 2L);
+            return MakeTupleFieldAccessAndReportUseSiteDiagnostics(tuple, tuple.Syntax, tuple.Type.TupleElements[i]);
+        }
+
+        private BoundExpression MakeTemp(BoundExpression expr, TupleBinaryOperatorInfo op, TupleOperatorSideEffectsAndTemps initialEffectsAndTemps, bool isRight)
+        {
+            if (op.IsSingle())
+            {
+                return EvaluateSideEffectingArgumentToTemp(VisitExpression(expr), isRight ? initialEffectsAndTemps.rightInit : initialEffectsAndTemps.leftInit, ref initialEffectsAndTemps.temps);
+            }
+            else
+            {
+                return expr;
+            }
+        }
+
+        /// <summary>
+        /// Produce an element-wise comparison and logic to ensure the result is a bool type.
+        /// 
+        /// If an element-wise comparison doesn't return bool, then:
+        /// - if it is dynamic, we'll do `!(comparisonResult.false)` or `comparisonResult.true`
+        /// - if it implicitly converts to bool, we'll just do the conversion
+        /// - otherwise, we'll do `!(comparisonResult.false)` or `comparisonResult.true` (as we'd do for `if` or `while`)
+        /// </summary>
+        private BoundExpression RewriteTupleSingleOperator(TupleBinaryOperatorInfo.Single single,
+            BoundExpression left, BoundExpression right, TypeSymbol boolType, BinaryOperatorKind operatorKind)
+        {
+            if (single.Kind.IsDynamic())
+            {
+                // Produce
+                // !((left == right).op_false)
+                // (left != right).op_true
+
+                var dynamicResult = _dynamicFactory.MakeDynamicBinaryOperator(single.Kind, left, right, isCompoundAssignment: false, _compilation.DynamicType).ToExpression();
+                if (operatorKind == BinaryOperatorKind.Equal)
+                {
+                    return _factory.Not(MakeUnaryOperator(UnaryOperatorKind.DynamicFalse, left.Syntax, method: null, dynamicResult, boolType));
+                }
+                else
+                {
+                    return MakeUnaryOperator(UnaryOperatorKind.DynamicTrue, left.Syntax, method: null, dynamicResult, boolType);
+                }
```

Comment:
Have we rewritten `left` and `right`?

Reply:
Yes. All the elements that arrive here come from `GetTuplePart` which produces either a `BoundLocal` or an `ItemN` access on a `BoundLocal`.
That said, I'll be removing the `VisitExpression` below as we discussed.

---
In reply to: [168001590](https://github.com/dotnet/roslyn/pull/23356#discussion_r168001590) [](ancestors = 168001590)